flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/MemorySegmentReader.java

```diff
@@ -0,0 +1,126 @@
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.MemorySegment;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+/** A class that reads data cached in memory. */
+@Internal
+class MemorySegmentReader<T> implements SegmentReader<T> {
+
+    /** The tool to deserialize bytes into records. */
+    private final TypeSerializer<T> serializer;
+
+    /** The wrapper view of the input stream of memory segments to be used in TypeSerializer API. */
+    private final DataInputView inputView;
+
+    /** The total number of records contained in the segments. */
+    private final int totalCount;
+
+    /** The number of records that have been read so far. */
+    private int count;
+
+    MemorySegmentReader(TypeSerializer<T> serializer, Segment segment, int startOffset)
+            throws IOException {
+        ManagedMemoryInputStream inputStream = new ManagedMemoryInputStream(segment.getCache());
+        this.inputView = new DataInputViewStreamWrapper(inputStream);
+        this.serializer = serializer;
+        this.totalCount = segment.getCount();
+        this.count = 0;
+
+        for (int ignored = 0; ignored < startOffset; ignored++) {
+            next();
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        return count < totalCount;
+    }
+
+    @Override
+    public T next() throws IOException {
+        T ret = serializer.deserialize(inputView);
+        count++;
+        return ret;
+    }
+
+    @Override
+    public void close() {}
+
+    /** An input stream subclass that reads bytes from memory segments. */
+    private static class ManagedMemoryInputStream extends InputStream {
+
+        /** The memory segments to read bytes from. */
+        private final List<MemorySegment> segments;
+
+        /** The index of the segment that is currently being read. */
+        private int segmentIndex;
+
+        /** The number of bytes that have been read from current segment so far. */
+        private int segmentOffset;
+
+        public ManagedMemoryInputStream(List<MemorySegment> segments) {
+            this.segments = segments;
+            this.segmentIndex = 0;
+            this.segmentOffset = 0;
+        }
+
+        @Override
+        public int read() throws IOException {
+            int ret = segments.get(segmentIndex).get(segmentOffset) & 0xff;
+            segmentOffset += 1;
+            if (segmentOffset >= segments.get(segmentIndex).size()) {
+                segmentIndex++;
+                segmentOffset = 0;
+            }
+            return ret;
+        }
+
+        @Override
+        public int read(byte[] b, int off, int len) throws IOException {
+            int readLen = 0;
+
+            while (len > 0 && segmentIndex < segments.size()) {
```

Comment:
nits: the following code is probably simpler since it does not need to explicitly write `break`.

```
while (len > 0) {
    int currentLen = Math.min(segments.get(segmentIndex).size() - segmentOffset, len);
    segments.get(segmentIndex).get(segmentOffset, b, off, currentLen);
    segmentOffset += currentLen;
    if (segmentOffset >= segments.get(segmentIndex).size()) {
        segmentIndex++;
        segmentOffset = 0;
    }

    readLen += currentLen;
    off += currentLen;
    len -= currentLen;
}
```